backend/liveness/liveness.go

```diff
@@ -152,26 +155,64 @@ func (t *SwitchSet) Alive(ctx context.Context, id string, ttl int64) error {
 	})
 }
 
+// BuryAndRevokeLease is similar to Bury() but will revoke the lease associated
+// with the switch if one exists.
+func (t *SwitchSet) BuryAndRevokeLease(ctx context.Context, id string) error {
+	key := path.Join(t.prefix, id)
+
+	// find lease ID that we will revoke later on
+	var leaseID clientv3.LeaseID
+	var getResp *clientv3.GetResponse
+	if err := kvc.Backoff(ctx).Retry(func(n int) (done bool, err error) {
+		getResp, err = t.client.Get(ctx, key, clientv3.WithIgnoreValue())
+		return kvc.RetryRequest(n, err)
+	}); err != nil {
+		return fmt.Errorf("switch retrieval failed: %s", err)
+	}
+	if len(getResp.Kvs) != 0 {
+		leaseID = clientv3.LeaseID(getResp.Kvs[0].Lease)
+	}
+
+	// bury the key as usual
+	if err := t.Bury(ctx, id); err != nil {
+		return err
+	}
+
+	// revoke the lease from earlier if it exists
+	if leaseID != 0 {
+		if err := kvc.Backoff(ctx).Retry(func(n int) (done bool, err error) {
+			_, err = t.client.Revoke(ctx, leaseID)
+			return kvc.RetryRequest(n, err)
+		}); err != nil {
+			return fmt.Errorf("error burying switch: %s", err)
+		}
+	}
```

Comment:
I have some concerns about this section. If the lease revokes on etcd's end, or on another backend, before this backend's Revoke request, then this would produce a retry of the operation. Perhaps the revocation should be optimistic, and not produce an error if the operation fails.

Reply:
Makes sense. Update incoming.